src/cast_core/src/AssertiveActor.cpp

```diff
@@ -0,0 +1,216 @@
+    mongocxx::database database;
+    DocumentGenerator expectedExpr;
+    DocumentGenerator actualExpr;
+    std::optional<std::string> message;
+    std::set<std::string> ignoreFields;
+
+    PhaseConfig(PhaseContext& phaseContext, mongocxx::database&& db, ActorId id)
+        : database{db},
+          expectedExpr{phaseContext["Expected"].to<DocumentGenerator>(phaseContext, id)},
+          actualExpr{phaseContext["Actual"].to<DocumentGenerator>(phaseContext, id)},
+          message{phaseContext["Message"].maybe<std::string>()} {
+        for (auto [k, fieldNode] : phaseContext["IgnoreFields"]) {
+            ignoreFields.insert(fieldNode.to<std::string>());
+        }
+    }
+};
+
+const std::string prefix(AssertiveActor::PhaseConfig* config) {
+    return "Assert" +(config->message ? " \"" + *(config->message) + "\" ": "") + ": ";
+}
+
+auto runCommandAndGetResult(AssertiveActor::PhaseConfig* config, bsoncxx::document::value& command) {
+    auto cmdView = command.view();
+
+    BOOST_LOG_TRIVIAL(info) << prefix(config) << bsoncxx::to_json(cmdView);
+    auto res = config->database.run_command(std::move(command));
+    auto resView = res.view();
+
+    return res;
+}
+
+auto getBatchFromCommandResult(const bsoncxx::document::view& resView) {
+    if (resView["ok"]) {
+        auto cursor = resView["cursor"].get_document().view();
+        auto firstBatch = cursor["firstBatch"].get_array().value;
+        return firstBatch;
+    }
+    BOOST_THROW_EXCEPTION(AssertFailed("Failed to run command."));
+}
+
+template <typename T>
+double convertNumericToDouble(const T& numeric) {
+    switch(numeric.type()) {
+        case bsoncxx::type::k_double:
+            return numeric.get_double();
+        case bsoncxx::type::k_int32:
+            return numeric.get_int32() * 1.0;
+        case bsoncxx::type::k_int64:
+            return numeric.get_int64() * 1.0;
+        default:
+            BOOST_THROW_EXCEPTION(AssertFailed("not a numeric"));
+    }
+}
+
+bool isNumeric(bsoncxx::type type) {
+    return (type == bsoncxx::type::k_double)
+        || (type == bsoncxx::type::k_int32)
+        || (type == bsoncxx::type::k_int64);
+}
+
+bool equalBSONDocs(AssertiveActor::PhaseConfig* config, const bsoncxx::document::view& expected, const bsoncxx::document::view& actual);
+bool equalBSONArrays(AssertiveActor::PhaseConfig* config, const bsoncxx::array::view& expected, const bsoncxx::array::view& actual);
+
+template <typename T>
+bool equal(AssertiveActor::PhaseConfig* config, const T& expectedVal, const T& actualVal) {
+    auto type = expectedVal.type();
+    if (isNumeric(type) && isNumeric(actualVal.type())) {
+        // Rounding may result in one value being a double while another is integral. Handle this case through double comparison.
+        auto expectedDouble = convertNumericToDouble<T>(expectedVal);
+        auto actualDouble = convertNumericToDouble<T>(actualVal);
+        // Accept a small difference in numeric types.
```

Comment:
Because of the double conversion? Or something else? Should the allowed delta be configurable?

Reply:
Deferring that works for me! Can we just make sure it's extremely loudly clear and documented in both the actor header and in an example doc workload that this happens? I'd hate for someone to be surprised by this behavior, and have an error slip through when two values should have matched exactly, since they didn't dig through the implementation cpp.

Reply:
Done! I added a note in the header about it, and there is already a comment in the actor implementation. I also added a note about it to the validate.yml files for scale 1.